torch/csrc/jit/codegen/cuda/kernel.cpp

```diff
@@ -25,58 +26,135 @@ static int ceilDiv(const int a, const int b) {
   return (a + b - 1) / b;
 }
 
+// IO data structure for kernel code;
+static auto typeinfo = R"(
+template<typename T, int N>
+struct Tensor {
+  T& operator[](int ind) {
+    return data[ind];
+  };
+
+  T* data;
+  int size[N];
+  int stride[N];
+};
+)";
+
+// include IO data structure for host code
+struct KernelArgumentHolder {
+  std::vector<void*> arguments;
+  std::vector<char> buffer;
+  void* buffer_ptr;
+
+  KernelArgumentHolder(size_t n_inputs, size_t n_dimension_per_tensor) {
+    arguments.reserve(n_inputs);
+    // We are being generous here on the allocated buffer;
+    buffer.resize(
+        n_inputs * (sizeof(void*) + 2 * sizeof(int) * n_dimension_per_tensor));
+    buffer_ptr = buffer.data();
+  }
+
+  void** args() {
+    return arguments.data();
+  }
+
+  // this should comply to the storage of Tensor object defined in typeinfo;
+  void push_tensor(
+      const at::Tensor& val,
+      c10::optional<at::IntArrayRef> broadcasted_size = c10::nullopt) {
+    arguments.push_back(buffer_ptr);
+
+    // passing address, type doesn't really matter here;
+    auto data_ptr = static_cast<void**>(buffer_ptr);
+    *data_ptr = val.data_ptr();
+    buffer_ptr += sizeof(char*);
```

Comment:
Maybe `buffer_ptr` should be local, not a member of the class. I get you're trying to scan it, but it seems the offset is simple if you know which argument you're working on. I'm worried if it's not called in the right order `buffer_ptr` could get desynced.

Reply:
Considering that we might have scalar inputs. I don't think we can easily index this.

In the case that graph inputs and generated kernel arguments are out of sync, we have bigger problem than mismatch arguments. We are basically generating wrong kernel in that situation.

Reply:
I was thinking you could divvy up the buffer at the start, record the section lengths, and have an map from argument index to offset. This way you don't have to carry this information from function to function. But this is more of a nit at this point.